### ISSUE TITLE
--no-restore broken

### ISSUE THREAD

**Issue by draganjovanovic1** [fsprojects/Paket#3505]
### Description

With 5.197.0 when run command `dotnet build --no-restore` restore is still being executed

```
Dragans-MBP:xyz dragan$ dotnet build --no-restore
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Paket version 5.197.0
  Paket version 5.197.0
  Paket version 5.197.0
  Paket version 5.197.0
  /Users/dragan/xyz/demo-data/import-tool/obj/paket.locked is locked by paket.exe (PID = 50316). Waiting...
  /Users/dragan/xyz/demo-data/import-tool/obj/paket.locked is locked by paket.exe (PID = 50316). Waiting...
  /Users/dragan/xyz/project-templates/webapi/content/obj/paket.locked is locked by paket.exe (PID = 50289). Waiting...
  /Users/dragan/xyz/project-templates/webapi/content/obj/paket.locked is locked by paket.exe (PID = 50289). Waiting...
  /Users/dragan/xyz/project-templates/webapi/content/obj/paket.locked is locked by paket.exe (PID = 50289). Waiting...
  /Users/dragan/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 50289). Waiting...
  /Users/dragan/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 50289). Waiting...
  /Users/dragan/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 50289). Waiting...
  /Users/dragan/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 50318). Waiting...
  /Users/dragan/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 50318). Waiting...
```

### Repro steps

Run `dotnet build --no-restore` on solution

### Expected behavior

Restore shouldn't be executed when `--no-restore` flag is provided

### Actual behavior

Restore is being executed

**Comment by forki** [comment]
did this work with older versions?

**Comment by draganjovanovic1** [comment]
Yes, it works with `5.196.2`

**Comment by forki** [comment]
I think even older versions actually do call paket and the change just surfaced that. Do you know which MSBuild property is set by --no-restore?

**Comment by draganjovanovic1** [comment]
If I remember correctly this was an issue with previous versions of paket as well but was fixed recently. I really don't know what MsBuild property is being set. Will try to take a look later today.

**Comment by forki** [comment]
I assume it's `ResolveNugetPackages`

**Comment by forki** [comment]
please retry with latest

**Comment by draganjovanovic1** [comment]
It is fixed now. Not doing restore. However, I am kind of confused here. Release `5.197.0` resolves the [issue](https://github.com/ionide/ionide-vscode-paket/issues/68) which says that `paket install` should avoid need to exec `dotnet restore` afterwards but it doesn't:

```
dotnet build --no-restore
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/xyz/src/backend/Core/Cqrs/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/xyz/src/backend/Core/Cqrs/Core.Cqrs.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/xyz/src/backend/Core/Logging/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/xyz/src/backend/Core/Logging/Core.Logging.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/xyz/src/backend/Core/Mq/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/xyz/src/backend/Core/Mq/Core.Mq.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/xyz/src/backend/Core/EsToMqConnector/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/xyz/src/backend/Core/EsToMqConnector/Core.EsToMqConnector.fsproj]
...
```

It is obvious that `paket install` is now way slower than before so it looks like it does some more work but `dotnet restore` is still required in order to compile project.

**Comment by forki** [comment]
well paket install doesn't create the assets file. That's done by dotnet restore. So you still need that. The fix in 5.197.0 is only referring to the VS Code / VS part which now can do background restore. It does not change behaviour for cmd line

**Comment by draganjovanovic1** [comment]
Ok, I misunderstood it. 

I tried this scenario: `paket install` then `dotnet build` (without no-restore flag) and it does not compile, reports bunch of errors with undefined types and namespaces so I had to revert to `5.196.2` for now as I have some urgent work to do and can't debug right now. 

The most annoying part with `paket` and `dotnet restore` is a bunch of `paket.locked is locked by paket.exe` errors where it hangs for significant amount of time and it makes work with NuGets really painful. Is there any known workaround for this?

**Comment by forki** [comment]
why do you do --no-restore? Nuget needs to create the assets file.

   `paket.locked is locked by paket.exe` errors 

that's not an error. it's just a point where we work sequentially. If we can reproduce it in a small repro then we might do more work in parallel

**Comment by forki** [comment]
seems this fix wasn't good. https://github.com/fsprojects/Paket/issues/3505

**Comment by forki** [comment]
ok had to revert. will ttry another fix as alpha prerelease later today

**Comment by BlythMeister** [comment]
I have previously fixed this to get it working.

The issue was that the props version being written did not match the version checked in the restore targets.

Have you got the latest targets?

**Comment by BlythMeister** [comment]
See #3449

**Comment by forki** [comment]
I beleive that's not the real fix for this issue here

**Comment by BlythMeister** [comment]
I've tried to repo this and can't...
Can the OP create an example repository which demonstrates this?

**Comment by forki** [comment]
@BlythMeister remove paket-files folder (and maybe obj) and run `dotnet build --no-restore` in one of the project folders. If you see paket output then it reproed

**Comment by BlythMeister** [comment]
--no-restore will call paket...unless you have already restored some other way.

The only time you should use --no-restore is if you have previously called dotnet restore.

**Comment by BlythMeister** [comment]
If --no-restore doesn't call paket...your build will fail... since none of the dependencies will be written into the obj JSON files.

**Comment by forki** [comment]
No it's actually fine. All paket files are already generated during paket
install. The only thing missing is the assets file that nuget generates.

So repro is: clear, paket install, dotnet build --no-restore

Am Di., 19. Feb. 2019, 18:04 hat Chris Blyth <notifications@github.com>
geschrieben:

**Comment by BlythMeister** [comment]
Ok, then question to the OP.

before calling dotnet build are you doing a dotnet restore, or paket install?
Since the repo in the OP doesn't state that.

**Comment by draganjovanovic1** [comment]
To give you more insight, I use have solution with more than 50 `netcoreapp2.2` and `netstandard2.0` projects written in F#. 

I just tested `5.197.2` and here are my findings.

## 1st try - unclear build output

Delete all `bin`, `obj`, `packages`, etc., to have clear situation:
```
Dragans-MacBook-Pro:xyz dragan$ find . -type d -name bin -exec rm -r {} \;
Dragans-MacBook-Pro:xyz dragan$ find . -type d -name obj -exec rm -r {} \;
Dragans-MacBook-Pro:xyz dragan$ rm -r packages/
Dragans-MacBook-Pro:xyz dragan$ rm -r .fake/
Dragans-MacBook-Pro:xyz dragan$ rm -r paket-files/
```

Do `paket install`:
```
Dragans-MacBook-Pro:xyz dragan$ mono .paket/paket.exe install

Paket version 5.197.2
Skipping resolver for group Main since it is already up-to-date
/Users/dragan/Projects/xyz/paket.lock is already up-to-date
Installing into projects:
 - Creating model and downloading packages.
Downloading chromedriver.storage.googleapis.com/:/2.43/chromedriver_linux64.zip linux64/chromedriver_linux64.zip to /Users/dragan/Projects/xyz/paket-files/chromedriver.storage.googleapis.com/linux64/chromedriver_linux64.zip
Downloading chromedriver.storage.googleapis.com/:/2.43/chromedriver_mac64.zip macos/chromedriver_mac64.zip to /Users/dragan/Projects/xyz/paket-files/chromedriver.storage.googleapis.com/macos/chromedriver_mac64.zip
Downloading chromedriver.storage.googleapis.com/:/2.43/chromedriver_win32.zip win32/chromedriver_win32.zip to /Users/dragan/Projects/xyz/paket-files/chromedriver.storage.googleapis.com/win32/chromedriver_win32.zip
Downloading download.microsoft.com/:/download/F/8/A/F8AF50AB-3C3A-4BC4-8773-DC27B32988DD/MicrosoftWebDriver.exe win32/MicrosoftWebDriver.exe to /Users/dragan/Projects/xyz/paket-files/download.microsoft.com/win32/MicrosoftWebDriver.exe
Downloading github.com/:/mozilla/geckodriver/releases/download/v0.23.0/geckodriver-v0.23.0-linux32.tar.gz linux32/geckodriver_linux32.tar.gz to /Users/dragan/Projects/xyz/paket-files/github.com/linux32/geckodriver_linux32.tar.gz
Downloading github.com/:/mozilla/geckodriver/releases/download/v0.23.0/geckodriver-v0.23.0-linux64.tar.gz linux64/geckodriver_linux64.tar.gz to /Users/dragan/Projects/xyz/paket-files/github.com/linux64/geckodriver_linux64.tar.gz
Downloading github.com/:/mozilla/geckodriver/releases/download/v0.23.0/geckodriver-v0.23.0-macos.tar.gz macos/geckodriver_macos.tar.gz to /Users/dragan/Projects/xyz/paket-files/github.com/macos/geckodriver_macos.tar.gz
Downloading github.com/:/mozilla/geckodriver/releases/download/v0.23.0/geckodriver-v0.23.0-win32.zip win32/geckodriver_win32.zip to /Users/dragan/Projects/xyz/paket-files/github.com/win32/geckodriver_win32.zip
Downloading github.com/:/mozilla/geckodriver/releases/download/v0.23.0/geckodriver-v0.23.0-win64.zip win64/geckodriver_win64.zip to /Users/dragan/Projects/xyz/paket-files/github.com/win64/geckodriver_win64.zip
 - Installing for projects
   - demo-data/import-tool/paket.references -> demo-data/import-tool/DemoData.fsproj
   - project-templates/webapi/content/paket.references -> project-templates/webapi/content/_AppName.fsproj
   
   ... AND LOT OF OTHER projects

Performance:
 - Disk IO: 10 seconds
 - Average Download Time: 3 seconds
 - Number of downloads: 9
 - Runtime: 5 minutes, 5 seconds

```

Then `dotnet build --no-restore`:
```
Dragans-MacBook-Pro:xyz dragan$ dotnet build --no-restore

Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  shasum: /Users/dragan/Projects/xyz/.paket/../paket-files/paket.restore.cached: 
  Paket version 5.197.2
  shasum: /Users/dragan/Projects/xyz/.paket/../paket-files/paket.restore.cached: 
  Paket version 5.197.2
  shasum: /Users/dragan/Projects/xyz/.paket/../paket-files/paket.restore.cached: 
  shasum: /Users/dragan/Projects/xyz/.paket/../paket-files/paket.restore.cached: 
  Paket version 5.197.2
  Paket version 5.197.2
  /Users/dragan/Projects/xyz/project-templates/webapi/content/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/demo-data/import-tool/obj/paket.locked is locked by paket.exe (PID = 78216). Waiting...
  /Users/dragan/Projects/xyz/project-templates/webapi/content/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/project-templates/webapi/content/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 78217). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 78217). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/EventPublisher/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/CommandHandler/obj/paket.locked is locked by paket.exe (PID = 78216). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/EventPublisher/obj/paket.locked is locked by paket.exe (PID = 78217). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Booking/EventPublisher/obj/paket.locked is locked by paket.exe (PID = 78216). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Crm/EventPublisher/obj/paket.locked is locked by paket.exe (PID = 78216). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Identity/Server/obj/paket.locked is locked by paket.exe (PID = 78188). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Identity/Server/obj/paket.locked is locked by paket.exe (PID = 78217). Waiting...
  /Users/dragan/Projects/xyz/src/backend/Identity/Server/obj/paket.locked is locked by paket.exe (PID = 78216). Waiting...
  /Users/dragan/Projects/xyz/src/backend/ProcessManagers/CustomerActivation/obj/paket.locked is locked by paket.exe (PID = 78217). Waiting...
  Starting full restore process.
  Performance:
   - Disk IO: 94 milliseconds
   - Runtime: 7 minutes, 33 seconds
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/src/backend/Core/Cqrs/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/src/backend/Core/Cqrs/Core.Cqrs.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/src/backend/Core/Logging/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/src/backend/Core/Logging/Core.Logging.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/src/backend/Core/Mq/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/src/backend/Core/Mq/Core.Mq.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/src/backend/Core/EsToMqConnector/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/src/backend/Core/EsToMqConnector/Core.EsToMqConnector.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/src/backend/Crm/Domain/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/src/backend/Crm/Domain/Crm.Domain.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/src/backend/Crm/QueryApi/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/src/backend/Crm/QueryApi/Crm.QueryApi.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/src/backend/Crm/EventPublisher/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/src/backend/Crm/EventPublisher/Crm.EventPublisher.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/test/backend/Test.Helpers/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/test/backend/Test.Helpers/Test.Helpers.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/test/backend/Crm.Tests/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/test/backend/Crm.Tests/Crm.Tests.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/test/backend/Core.Library.Tests/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/test/backend/Core.Library.Tests/Core.Library.Tests.fsproj]
/usr/local/share/dotnet/sdk/2.2.104/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(208,5): error NETSDK1004: Assets file '/Users/dragan/Projects/xyz/test/backend/Core.ValueTypes.Tests/obj/project.assets.json' not found. Run a NuGet package restore to generate this file. [/Users/dragan/Projects/xyz/test/backend/Core.ValueTypes.Tests/Core.ValueTypes.Tests.fsproj]

... AND LOT OF THE SAME ERRORS that obj/project.assets.json file is not found 
```

The errors above are expected when `dotnet restore` wasn't executed prior to build with `--no-restore` flag. What is confuses me here is output before errors started to appear:
```
...paket.locked is locked by paket.exe (PID = 78217). Waiting...
  Starting full restore process.
```

## 2nd try - WORKS FINE

Delete all `bin`, `obj`, `packages`, etc., to have clear situation and do `paket install` with the same output as above.

Then `dotnet restore`:
```
  Restoring packages for /Users/dragan/Projects/xyz/src/backend/Booking/CommandHandler/Booking.CommandHandler.fsproj...
  Restoring packages for /Users/dragan/Projects/xyz/src/backend/Booking/QueryApi/Booking.QueryApi.fsproj...
  Restoring packages for /Users/dragan/Projects/xyz/src/backend/Booking/EventPublisher/Booking.EventPublisher.fsproj...
  Restoring packages for /Users/dragan/Projects/xyz/src/backend/Booking/Domain/Booking.Domain.fsproj...
  Generating MSBuild file /Users/dragan/Projects/xyz/src/backend/Booking/Domain/obj/Booking.Domain.fsproj.nuget.g.props.
  Generating MSBuild file /Users/dragan/Projects/xyz/src/backend/Booking/Domain/obj/Booking.Domain.fsproj.nuget.g.targets.
  Restore completed in 1.23 sec for /Users/dragan/Projects/xyz/src/backend/Booking/Domain/Booking.Domain.fsproj.

... AND LOT OF OTHER PROJECTS... ALL SUCCESSFULL
```

Then `dotnet build --no-restore`:
```
Dragans-MacBook-Pro:xyz dragan$ dotnet build --no-restore
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

...

Build succeeded.

Time Elapsed 00:04:19.27
```

**Comment by forki** [comment]
Ok that actually helps.

So there are at least 3 issues here.

1) we do not respect --no-restore. @enricosada is working on that
2) paket install is still not doing what paket restore does and that leads
to additional restores when it's actually not needed (independent of the
--no-restore flag) /cc @isaacabraham
3) parallel restore is blocking each other even in no-op cases


Am Di., 19. Feb. 2019, 19:55 hat Dragan Jovanović <notifications@github.com>
geschrieben:

**Comment by forki** [comment]
ok I found root casue for 2) - see https://github.com/fsprojects/Paket/issues/3508

**Comment by forki** [comment]
@draganjovanovic1 can you please try 5.197.3?

**Comment by BlythMeister** [comment]
I might be able to explain number 1...if you call dotnet build --no-restore but have not fully restored, a paket restore will trigger.  So your fix for 2 might also fix 1...

**Comment by BlythMeister** [comment]
I also see that parallel restore blocking as an issue in visual studio builds...it seems to do a full restore for every project, rather than just restore that 1 project.

**Comment by forki** [comment]
@BlythMeister 

1) is still an issue. we hook into the wrong targets
3) needs a nice repro with like 10 copied projects and simple command line calls.

**Comment by BlythMeister** [comment]
I can try

**Comment by BlythMeister** [comment]
@forki repo for 3) raised as new issue #3510

**Comment by draganjovanovic1** [comment]
@forki, will try `5.197.3` soon and post my findings.

**Comment by draganjovanovic1** [comment]
Ok, I tried the same 2 scenarios on the same solution.

1st scenario 

Works better now I would say. No more locks and restore messages in the output. 

Paket does not do `dotnet restore` which is probably expected so we have to do `dotnet restore` then `dotnet build --no-restore` (or `dotnet build` only) after `paket install`

2nd scenario 

Worked as before.

**Comment by draganjovanovic1** [comment]
I have tested `5.198.0` today and also works as expected. Closing
